contracts/PoaOperator.sol

```diff
@@ -291,6 +304,126 @@ contract PoaOperator is Adminable {
     bridge.deletePeriod(_period);
   }
 
+  // openTime could be derived from openPeriodHash's timestamp, but one has
+  // to determine what happens in the case that there was a long  (~ timeoutTime) time since
+  // the last submitted period, than it is possible to  open and timeout a challenge straight
+  // away
+  struct BeatChallenge {
+    address payable challenger;
+    uint256 openTime;
+    bytes32 openPeriodHash;
+  }
+
+  uint256 public minimumPulse; // max amount of periods one can go without a heartbeat
+  uint16 public heartbeatColor;
+  mapping(address => BeatChallenge) public beatChallenges;
+  
+  // challenger claims that there is no hearbeat included between (periodTime - brainDamageDuration) and periodTime
+  // TODO: figure out what happens in slot rotation
+  // TODO: must be sure that the surrent slot signer was also signer at (periodTime - brainDamageDuration)
+  function challengeBeat(
+    uint256 _slotId
+  ) public payable {
+    // check the stake
+    require(msg.value == vault.exitStake(), "invalid challenge stake");
+    
+    // check slot exists
+    require(_slotId < epochLength, "slotId too high");
+
+    // get the offending slot
+    Slot memory slot = slots[_slotId];
+    require(isSlotActive(slot), "Slot must be active");
+
+    bytes32 tip = bridge.tipHash();
+    
+    // check that challenge doesn't exist yet
+    // TODO: the slot signer can challenge himself, blocking further challenges for challengeTimeout duration. What are the implications of this?
+    require(beatChallenges[slot.signer].openTime == 0, "challenge already in progress");
+
+    // create challenge object
+    // TODO: only one challenge per address possible, potential problem if an address holds multiple slots 
+    beatChallenges[slot.signer] = BeatChallenge({
+      challenger: msg.sender,
+      openTime: now,
+      openPeriodHash: tip
+    });
+  }
+
+  // In case of an invalid challenge, can submit a proof of heartbeat
+  // TODO: Is the signer we challenged still in control of the Slot? Depends on timeout-time (currently casChallangeDuration) at the least.
+  function respondBeat(
+    bytes32[] memory _inclusionProof,
+    bytes32[] memory _walkProof,
```

Comment:
@TheReturnOfJan could you please more comments about what is `walkProof` ?

Reply:
@troggy new commit should be a bit clearer.